parlai/tasks/wizard_of_internet/mutators.py

```diff
@@ -0,0 +1,255 @@
+                for i in range(len(checked_sentences)):
+                    d = d.replace(
+                        "||CHECKED_SENTENCE_" + str(i) + "||", checked_sentences[i]
+                    )
+                new_docs.append(d)
+                new_titles.append(titles[ind])
+                new_urls.append(urls[ind])
+                break
+            else:
+                new_d = d[0:end_chunk]
+                for i in range(len(checked_sentences)):
+                    new_d = new_d.replace(
+                        "||CHECKED_SENTENCE_" + str(i) + "||", checked_sentences[i]
+                    )
+                new_docs.append(new_d)
+                new_titles.append(titles[ind])
+                new_urls.append(urls[ind])
+                d = d[end_chunk + 1 : -1]
+    new_message.force_set(CONST.RETRIEVED_DOCS, new_docs)
+    new_message.force_set(CONST.RETRIEVED_DOCS_TITLES, new_titles)
+    new_message.force_set(CONST.RETRIEVED_DOCS_URLS, new_urls)
+    return new_message
+
+
+@register_mutator("woi_chunk_retrieved_docs")
+class WoiChunkRetrievedDocs(MessageMutator):
+    """
+    Chunks '__retrieved-docs__' into smaller docs (max 100 words each).
+    """
+
+    @classmethod
+    def add_cmdline_args(
+        cls, parser: ParlaiParser, partial_opt: Optional[Opt] = None
+    ) -> ParlaiParser:
+        parser.add_argument(
+            '--woi-doc-chunk-size',
+            default=500,
+            type=int,
+            help='Document chunk size (in characters).',
+        )
+
+    def message_mutation(self, message: Message) -> Message:
+        chunk_sz = self.opt.get('woi_doc_chunk_size', 500)
+        return chunk_docs_in_message(message, chunk_sz)
+
+
+@register_mutator("woi_dropout_retrieved_docs")
+class WoiDropoutRetrievedDocs(MessageMutator):
+    """
+    Drops out '__retrieved-docs__' to only keep a maximum number in each example.
+    """
+
+    @classmethod
+    def add_cmdline_args(
+        cls, parser: ParlaiParser, partial_opt: Optional[Opt] = None
+    ) -> ParlaiParser:
+        parser.add_argument(
+            '--woi-doc-max-chunks',
+            default=100,
+            type=int,
+            help='Largest number of chunks to use, others will be dropped out at random. Chunks containing gold checked sentences will not be removed.',
+        )
+
+    def message_mutation(self, message: Message) -> Message:
+        if CONST.RETRIEVED_DOCS not in message:
+            return message
+        new_message = message.copy()
+        docs = message.get(CONST.RETRIEVED_DOCS)
+        new_docs = []
+        max_chunks = self.opt.get('woi_doc_max_chunks', 100)
+
+        keep = torch.randperm(len(docs))[0:max_chunks]
+        remove = torch.ones(len(docs))
+        remove[keep] = 0
+
+        for i in range(len(docs)):
+            if remove[i] == 0:
+                new_docs.append(docs[i])
+            else:
+                # We may still keep the doc if it contains the gold checked sentence(s).
```

Comment:
In that case should we mark one of the other docs to be removed instead?

Reply:
we'll just have one less, i dont think it's a big deal